core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -138,76 +155,106 @@ void setDelegate(LoadBalancer lb) {
+      if (s.getAttributes().get(GrpcAttributes.ATTR_LB_ADDR_AUTHORITY) != null) {
+        haveBalancerAddress = true;
+        break;
       }
+    }
 
-      if (haveBalancerAddress) {
-        return getProviderOrThrow("grpclb", "NameResolver has returned balancer addresses");
-      }
+    if (haveBalancerAddress) {
+      LoadBalancerProvider provider =
+          getProviderOrThrow("grpclb", "NameResolver has returned balancer addresses");
+      return new PolicySelection(provider, null);
+    }
 
-      String serviceConfigChoiceBalancingPolicy = null;
-      if (config != null) {
-        serviceConfigChoiceBalancingPolicy =
-            ServiceConfigUtil.getLoadBalancingPolicyFromServiceConfig(config);
-        if (serviceConfigChoiceBalancingPolicy != null) {
-          // Handle ASCII specifically rather than relying on the implicit default locale of the str
-          return getProviderOrThrow(
-              Ascii.toLowerCase(serviceConfigChoiceBalancingPolicy),
-              "service-config specifies load-balancing policy");
+    if (config != null) {
+      List<Map<String, Object>> lbConfigs =
+          ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
+      LinkedHashSet<String> policiesTried = new LinkedHashSet<String>();
+      for (Map<String, Object> lbConfig : lbConfigs) {
+        if (lbConfig.size() != 1) {
+          throw new PolicyException(
+              "There are " + lbConfig.size()
+              + " load-balancing configs in a list item. Exactly one is expected. Config="
+              + lbConfig);
         }
+        Entry<String, Object> entry = lbConfig.entrySet().iterator().next();
+        // Internally the service config is written in Protobuf and converted to JSON.
+        // Unfortunately the policy name is lowerCamel format in the converted JSON.  We will
+        // have to convert it to the lower_underscore format to match the policy names.
+        // The conversion is no-op if it's already lower_underscore, thus external users who
+        // write the service config in JSON directly can use lower_underscore.
+        String policy = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, entry.getKey());
+        LoadBalancerProvider provider = registry.getProvider(policy);
+        if (provider != null) {
+          if (!policiesTried.isEmpty()) {
+            channelLogger.log(
+                ChannelLogLevel.DEBUG,
+                "{0} specified by Service Config are not available", policiesTried);
+          }
+          return new PolicySelection(provider, (Map) entry.getValue());
+        }
+        policiesTried.add(policy);
       }
-      return getProviderOrThrow(DEFAULT_POLICY, "Using default policy");
+      throw new PolicyException(
+          "None of " + policiesTried + " specified by Service Config are available.");
     }
+    return new PolicySelection(
+        getProviderOrThrow(DEFAULT_POLICY, "Using default policy"), null);
```

Comment:
Given how the string is combined, it seems this should be s/U/u/ or the string should be quoted.

Reply:
done.